lib/commands/publish/publishChangedPackages.js

```diff
@@ -3,33 +3,116 @@ var checkUpdatedPackages  = require("../../utils/checkUpdatedPackages");
 var npmPublishSemiAtomic  = require("./npmPublishSemiAtomic");
 var packageUtils          = require("../../utils/packageUtils");
 var unionWith             = require("lodash.unionwith");
-var readline              = require("readline-sync");
 var gitUtils              = require("../../utils/gitUtils");
+var inquirer              = require("inquirer");
 var fsUtils               = require("../../utils/fsUtils");
 var logger                = require("../../utils/logger");
 var semver                = require("semver");
+var async                 = require("async");
+
+function promptVersion(currentVersion, pkgName, callback) {
+  var patch = semver.inc(currentVersion, "patch");
+  var minor = semver.inc(currentVersion, "minor");
+  var major = semver.inc(currentVersion, "major");
+
+  var message = "Select a new version";
+  if (pkgName) message += " for " + pkgName;
+  message += " (currently " + currentVersion + ")";
+
+  inquirer.prompt([{
+    type: "list",
+    name: "version",
+    message: message,
+    default: false,
+    choices: [
+      { value: patch, name: "Patch (" + patch + ")" },
+      { value: minor, name: "Minor (" + minor + ")" },
+      { value: major, name: "Major (" + major + ")" },
+      { value: false, name: "Custom" }
+    ]
+  }, {
+    type: "input",
+    name: "custom",
+    message: "Enter a custom version",
+    filter: semver.valid,
+    validate: function (version) {
+      return semver.valid(version) ? true : "Must be a valid semver version";
+    },
+    when: function (answers) {
+      return !answers.version; // Only when "Custom"
+    }
+  }], function (answers) {
+    callback(null, answers.version || answers.custom);
+  });
+}
+
+function getVersions(changedPackages, globalVersion, flags, callback) {
+  var versions = {};
 
-function userGetVersion(currentVersion, pkgName) {
-  var message = "New version ";
-  if (pkgName) message += "for \"" + pkgName + "\" ";
-  message += "(Leave blank for patch version): ";
+  if (flags.canary) {
+    var version = globalVersion + "-canary." + gitUtils.getCurrentSHA().slice(0, 8);
 
-  var input = readline.question(message);
+    changedPackages.forEach(function (pkg) {
+      versions[pkg.name] = version;
+    });
 
-  var ver = semver.valid(input);
+    callback(null, {
+      version: version,
+      versions: versions
+    });
+  } else if (!flags.independent) {
```

Comment:
Trying to group together canary and non-independent mode since they are similar to one another